loopy/transform/reindexing.py

```diff
@@ -0,0 +1,339 @@
+
+    for insn in subst_kernel.instructions:
+        domain = subst_kernel.get_inames_domain(insn.within_inames)
+        if isinstance(insn, MultiAssignmentBase):
+            access_exprs = _IndexCollector(var_name)((insn.assignees,
+                                                      insn.expression,
+                                                      tuple(insn.predicates)))
+        elif isinstance(insn, (CInstruction, BarrierInstruction)):
+            access_exprs = _IndexCollector(var_name)(insn.predicates)
+        else:
+            raise NotImplementedError(type(insn))
+
+        for access_expr in access_exprs:
+            access_ranges.append(get_access_map(domain,
+                                                access_expr.index_tuple,
+                                                assumptions=kernel.assumptions)
+                                 .range())
+
+    del subst_kernel
+
+    # }}}
+
+    access_range = _union_access_ranges(access_ranges)
+
+    subst, new_shape = _get_seghir_loechner_reindexing_from_range(
+        access_range)
+
+    # {{{ simplify new_shape with the assumptions from kernel
+
+    new_shape = new_shape.gist_params(kernel.assumptions)
+
+    # }}}
+
+    # {{{  update kernel.temporary_variables
+
+    new_shape = new_shape.drop_unused_params()
+
+    new_temps = dict(kernel.temporary_variables).copy()
+    new_temps[new_var_name] = new_temps.pop(var_name).copy(
+        name=new_var_name,
+        shape=pw_qpolynomial_to_expr(new_shape),
+        strides=None,
+        dim_tags=None,
+        dim_names=None,
+    )
+
+    kernel = kernel.copy(temporary_variables=new_temps)
+
+    # }}}
+
+    # {{{ perform the substitution i.e. reindex the accesses
+
+    subst_expr = pw_qpolynomial_to_expr(subst)
+    subst_dim_names = tuple(
+        subst.space.get_dim_name(isl.dim_type.param, idim)
+        for idim in range(access_range.dim(isl.dim_type.out)))
+    assert not (set(subst_dim_names) & kernel.all_variable_names())
+
+    rule_mapping_context = SubstitutionRuleMappingContext(kernel.substitutions,
+                                                          vng)
+    reindexing_mapper = ReindexingApplier(rule_mapping_context,
+                                          var_name, new_var_name,
+                                          subst_expr, subst_dim_names)
+
+    def _does_access_var_name(kernel, insn, *args):
+        return var_name in insn.dependency_names()
+
+    kernel = reindexing_mapper.map_kernel(kernel,
+                                          within=_does_access_var_name,
+                                          map_args=False,
+                                          map_tvs=False)
+    kernel = rule_mapping_context.finish_kernel(kernel)
+
+    # }}}
+
+    if not i_agree_to_the_gpl_terms:
+        from warnings import warn
+        warn("The returned kernel object is a derivative of "
+             "Barvinok library, making it a derivative of "
```

Comment:
Fix the phrasing here, too. Arguably, a warning is too easy to ignore. I think this should be an error.

Reply:
Dropped this piece of code as discussed in https://github.com/inducer/loopy/pull/636#discussion_r914414855.